packages/terra-date-input/src/DateInput.jsx

```diff
@@ -0,0 +1,662 @@
+      year: DateInputUtil.splitYear(value),
+      isFocused: false,
+      monthIsFocused: false,
+      dayIsFocused: false,
+      yearIsFocused: false,
+    };
+  }
+
+  componentDidUpdate(prevProps) {
+    if (this.props.value === prevProps.value) {
+      return;
+    }
+
+    // eslint-disable-next-line react/no-did-update-set-state
+    this.setState({
+      month: DateInputUtil.splitMonth(this.props.value),
+      day: DateInputUtil.splitDay(this.props.value),
+      year: DateInputUtil.splitYear(this.props.value),
+    });
+  }
+
+  handleFocus(event) {
+    if (this.props.onFocus && !this.dateInputContainer.current.contains(event.relatedTarget)) {
+      this.props.onFocus(event);
+    }
+
+    this.setState({ isFocused: true });
+  }
+
+  handleMonthFocus(event) {
+    this.handleFocus(event);
+    this.setState({ monthIsFocused: true });
+  }
+
+  handleDayFocus(event) {
+    this.handleFocus(event);
+    this.setState({ dayIsFocused: true });
+  }
+
+  handleYearFocus(event) {
+    this.handleFocus(event);
+    this.setState({ yearIsFocused: true });
+  }
+
+  handleMonthBlur(event) {
+    this.handleBlur(event, DateInputUtil.inputType.MONTH);
+    this.setState({ monthIsFocused: false });
+  }
+
+  handleDayBlur(event) {
+    this.handleBlur(event, DateInputUtil.inputType.DAY);
+    this.setState({ dayIsFocused: false });
+  }
+
+  handleYearBlur(event) {
+    this.handleBlur(event, DateInputUtil.inputType.YEAR);
+    this.setState({ yearIsFocused: false });
+  }
+
+  handleBlur(event, type) {
+    this.setState({ isFocused: false });
+
+    if (type === DateInputUtil.inputType.DAY) {
+      let stateValue = event.target.value;
+
+      // Prepend a 0 to the value when losing focus and the value is single digit.
+      if (stateValue.length === 1) {
+        stateValue = '0'.concat(stateValue);
+
+        this.handleValueChange(event, type, stateValue);
+      }
+    }
+
+    if (this.props.onBlur) {
+      // Modern browsers support event.relatedTarget but event.relatedTarget returns null in IE 10 / IE 11.
+      // IE 11 sets document.activeElement to the next focused element before the blur event is called.
+      const activeTarget = event.relatedTarget ? event.relatedTarget : document.activeElement;
+
+      // Handle blur only if focus has moved out of the entire date input component.
+      if (!this.dateInputContainer.current.contains(activeTarget)) {
+        this.props.onBlur(event);
+      }
+    }
+  }
+
+  handleMonthChange(event) {
+    if (!DateInputUtil.validNumericInput(event.target.value)) {
+      return;
+    }
+
+    const inputValue = event.target.value;
+    const stateValue = this.state.month;
+    const maxValue = 12;
+
+    // Ignore the entry if the value did not change or it is invalid.
+    if (inputValue === stateValue || Number(inputValue) > maxValue) {
+      return;
+    }
+
+    this.handleValueChange(event, DateInputUtil.inputType.MONTH, inputValue);
+  }
+
+  handleDayChange(event) {
```

Comment:
I had coded in the auto-focus shift functionality whenever month / day values were filled in but ran into an issue when using only a keyboard to set the month value in the month select.

For example, if I type the character `J` while focused on the month input, it will switch the month to January, updating the value from emptry string to `01`. The month select value is now populated and so the auto-focus logic would kick in shifting the focus to the day input in the `month-day-year` formatted date input. 

This caused an issue as the user may not have wanted to select January when they typed in the letter `J` and maybe they were trying to type in June or July. I had considered adding a delay before auto-shifting focus from the month so they could type in `Jun` but thought introducing a delay on the month select would be weird. And only auto-shifting focus on the day input also felt like a weird possibility so I decided to remove auto-focus shift functionality instead.

Reply:
Thanks for the clarification. We can keep it as is and enhance it when we figure out a good way to handle it.

Reply:
Talked with Ben offline about possibly enabling auto-focus if the user is selecting a month by not typing in the name (the user clicks on the month select and then selects a month or the user press arrow keys up/down to select a month and then press enter) as a future enhancement if we think this is functionality that would be good to add.